snakemake/executors.py

```diff
@@ -2168,3 +2171,878 @@ def run_wrapper(
+            """A helper function to check the filesize, and return the file
+               to the calling function
+            """
+            gb = bytesto(os.stat(f).st_size, "g")
+            if gb > warning_size_gb:
+                logger.warning(
+                    "File {} (size {} GB) is greater than the {} GB suggested size "
+                    "Consider uploading larger files to storage first.".format(
+                        f, gb, warning_size_gb
+                    )
+                )
+            return f
+
+        self.workflow_sources = []
+
+        for wfs in self.workflow.get_sources():
+            if os.path.isdir(wfs):
+                for (dirpath, dirnames, filenames) in os.walk(wfs):
+                    self.workflow_sources.extend(
+                        [check_size(os.path.join(dirpath, f)) for f in filenames]
+                    )
+            else:
+                self.workflow_sources.append(check_size(os.path.abspath(wfs)))
+
+    def _generate_build_source_package(self):
+        """in order for the instance to access the working directory in storage,
+           we need to upload it. This file is cleaned up at the end of the run.
+           We do this, and then obtain from the instance and extract.
+        """
+        # Workflow sources for cloud executor must all be under same workdir root
+        for filename in self.workflow_sources:
+            if self.workdir not in filename:
+                raise WorkflowError(
+                    "All source files must be present in the working directory, "
+                    "{workdir} to be uploaded to a build package that respects "
+                    "relative paths, but {filename} was found outside of this "
+                    "directory. Please set your working directory accordingly, "
+                    "and the path of your Snakefile to be relative to it.".format(
+                        workdir=self.workdir, filename=filename
+                    )
+                )
+
+        # We will generate a tar.gz package, renamed by hash
+        tmpname = next(tempfile._get_candidate_names())
+        targz = os.path.join(tempfile.gettempdir(), "snakemake-%s.tar.gz" % tmpname)
+        tar = tarfile.open(targz, "w:gz")
+
+        # Add all workflow_sources files
+        for filename in self.workflow_sources:
+            arcname = filename.replace(self.workdir + os.path.sep, "")
+            tar.add(filename, arcname=arcname)
+
+        tar.close()
+
+        # Rename based on hash, in case user wants to save cache
+        sha256 = get_file_hash(targz)
+        hash_tar = os.path.join(self.workdir, "snakeworkdir-%s.tar.gz" % sha256)
+
+        # Only copy if we don't have it yet, clean up if we do
+        if not os.path.exists(hash_tar):
+            shutil.move(targz, hash_tar)
+        else:
+            os.remove(targz)
+
+        # We will clean these all up at shutdown
+        self._build_packages.add(hash_tar)
```

Comment:
There should only be one archive generated. It will be the same for all jobs. Hence, this does not need to be a set.

Reply:
We've had this discussion before and I've tried it, and it doesn't work. See https://github.com/snakemake/snakemake/pull/90#pullrequestreview-338887510.

Reply:
Mhm, in that comment you said that it is fixed. The job step (or jobscript) does not need to be part of the archive. It does not even need to be a script. See the kubernetes executor, where we just specify a shell command to be run. The job properties line you refer to in your previous [comment](https://github.com/snakemake/snakemake/pull/90#pullrequestreview-338887510) is meaningless here, it is for cluster environments. Just directly specify the exec statement when submitting the job.

Reply:
@johanneskoester I think we already switched it so that the exec_job statement is the command given to the container, e.g.,
```python
        exec_job = self.format_job(
            self.exec_job, job, _quote_all=True, use_threads=use_threads
        )

        # Now that we've parsed the job resource requirements, add to exec
        exec_job += self.get_default_resources_args()

        # The full command to download the archive, extract, and run
        # For snakemake bases, we must activate the conda environment, but
        # for custom images we must allow this to fail (hence || true)
        commands = [
            "/bin/bash",
            "-c",
            "mkdir -p /workdir && cd /workdir && wget -O /download.py https://gist.githubusercontent.com/vsoch/84886ef6469bedeeb9a79a4eb7aec0d1/raw/181499f8f17163dcb2f89822079938cbfbd258cc/download.py && chmod +x /download.py && source activate snakemake || true && pip install crc32c && python /download.py download %s %s /tmp/workdir.tar.gz && tar -xzvf /tmp/workdir.tar.gz && %s"
            % (self.bucket.name, self.pipeline_package, exec_job),
        ]
```
and if we look at logs [from the tutorial](https://github.com/snakemake/snakemake/pull/331/files#diff-eb71c39d80e149671e6f4b862fda5612R730) we can verify that is what is happening. So I'll test moving the archive generation to the beginning again, it will either work or there is something else I'm forgetting :)

Reply:
Testing now!

Reply:
okay that worked! Sorry for not trying it before remembering that we made all these changes.

Reply:
Good work, thanks!